### PR TITLE
docs: Add more detailed architecture documentation

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

Add documentation to make it easier for new contributors to understand the architecture.

### CODE REVIEW

**Comment by andygrove** [comment]
This is ready for review now @viirya @comphead

File: docs/source/contributor-guide/plugin_overview.md [hunk 1]
```diff
@@ -17,30 +17,37 @@ specific language governing permissions and limitations
...
+The entry point to Comet is the `org.apache.spark.CometPlugin` class, which can be registered with Spark by adding the 
+following setting to the Spark configuration when launching `spark-shell` or `spark-submit`:
 
 ```
 --conf spark.plugins=org.apache.spark.CometPlugin
 ```
 
-On initialization, this class registers two physical plan optimization rules with Spark: `CometScanRule` and `CometExecRule`. These rules run whenever a query stage is being planned.
+On initialization, this class registers two physical plan optimization rules with Spark: `CometScanRule` 
+and `CometExecRule`. These rules run whenever a query stage is being planned during Adaptive Query Execution.
```
Comment (comphead) [inline docs/source/contributor-guide/plugin_overview.md:32]: does that mean CometExecRule is not called if AQE is disabled?

File: docs/source/contributor-guide/plugin_overview.md [hunk 2]
```diff
@@ -17,30 +17,37 @@ specific language governing permissions and limitations
...
+`CometScanRule` replaces any Parquet scans with Comet operators. There are different paths for v1 and v2 data sources.
 
-When the V1 data source API is being used, `FileSourceScanExec` is replaced with `CometScanExec`.
+When reading from Parquet v1 data sources, Comet replaces `FileSourceScanExec` with a `CometScanExec`, and for v2 
+data sources, `BatchScanExec` is replaced with `CometBatchScanExec`. In both cases, Comet replaces Spark's Parquet 
+reader with a custom vectorized Parquet reader. This is similar to Spark's vectorized Parquet reader used by the v2 
+Parquet data source but leverages native code for decoding Parquet row groups directly into Arrow format.
 
-When the V2 data source API is being used, `BatchScanExec` is replaced with `CometBatchScanExec`.
+Comet only supports a subset of data types and will fall back to Spark's scan if unsupported types
```
Comment (comphead) [inline docs/source/contributor-guide/plugin_overview.md:43]: should we name the types or make a link to the code what types are supported?

**Review by comphead — APPROVED** [review]
lgtm thanks @andygrove

File: docs/source/contributor-guide/plugin_overview.md (hunk 1 above)
Comment (andygrove) [reply docs/source/contributor-guide/plugin_overview.md:32]: No, I should clarify that

File: docs/source/contributor-guide/plugin_overview.md (hunk 2 above)
Comment (andygrove) [reply docs/source/contributor-guide/plugin_overview.md:43]: Good point. I will link to https://datafusion.apache.org/comet/user-guide/datatypes.html

File: docs/source/contributor-guide/plugin_overview.md [hunk 3]
```diff
@@ -17,30 +17,41 @@ specific language governing permissions and limitations
...
 ```
 
-On initialization, this class registers two physical plan optimization rules with Spark: `CometScanRule` and `CometExecRule`. These rules run whenever a query stage is being planned.
+On initialization, this class registers two physical plan optimization rules with Spark: `CometScanRule`
+and `CometExecRule`. These rules run whenever a query stage is being planned during Adaptive Query Execution, and
+run once for the entire plan when Adaptive Query Execution is disabled.
 
 ## CometScanRule
 
-`CometScanRule` replaces any Parquet scans with Comet Parquet scan classes.
```
Comment (mbutrovich) [inline docs/source/contributor-guide/plugin_overview.md:32]: Maybe "Spark v1 and v2 data sources" to disambiguate that it's not a Parquet or Comet concept.

File: docs/source/contributor-guide/plugin_overview.md [hunk 4]
```diff
@@ -51,9 +62,27 @@ Comet does not support partially replacing subsets of the plan within a query st
...
+Once the plan has been transformed, any consecutive Comet operators are combined into a `CometNativeExec` which contains
+a serialized version of the plan (the serialization code can be found in `QueryPlanSerde`). When this operator is
+executed, the serialized plan is passed to the native code when calling `Native.createPlan`.
 
 In the native code there is a `PhysicalPlanner` struct (in `planner.rs`) which converts the serialized plan into an
-Apache DataFusion physical plan. In some cases, Comet provides specialized physical operators and expressions to
+Apache DataFusion `ExecutionPlan`. In some cases, Comet provides specialized physical operators and expressions to
 override the DataFusion versions to ensure compatibility with Apache Spark.
+
+`CometExecIterator` will invoke `Native.executePlan` to fetch the next batch from the native plan. This is repeated
```
Comment (mbutrovich) [inline docs/source/contributor-guide/plugin_overview.md:75]: fetch -> pull

**Review by mbutrovich — COMMENTED** [review]
Figures look great! Thanks for writing this up---it clears up a lot.

**Comment by andygrove** [comment]
Thanks for the reviews @comphead @mbutrovich

**Pull request merged**